### ISSUE TITLE
Implement bayesian optimisation of ml hyperparameters

### ISSUE THREAD

**Issue by marek-cottingham** [L-Sva/Team7_Code#5]
We can use the package https://github.com/fmfn/BayesianOptimization in order implement Bayesian optimisation. 
We should use the logs functionality to save progress as this could take some time for our model (retraining many times).

The black box function being optimised should take as input the model hyper-parameters and output the SB metric score for the best threshold value. All ML models generated should be saved between runs (ie. immediately before the black box function returns).

**Comment by marek-cottingham** [comment]
Candidate hyper-parameters to be optimised include:

- `n_estimators`: specifies the number of decision trees to be boosted. If n_estimator = 1, it means only 1 tree is generated, thus no boosting is at work. The default value is 100, but you can play with this number for optimal performance.
- `subsample`: it represents the subsample ratio of the training sample. A subsample = 0.5 means that 50% of training data is used prior to growing a tree. The value can be any fraction but the default value is 1.
- `max_depth`: it limits how deep each tree can grow. The default value is 6 but you can try other values if overfitting is an issue in your model.
- `learning_rate` (alias: eta): it is a regularization parameter that shrinks feature weights in each boosting step. The default value is 0.3 but people generally tune with values such as 0.01, 0.1, 0.2 etc.
- `gamma` (alias: min_split_loss): it’s another regularization parameter for tree pruning. It specifies the minimum loss reduction required to grow a tree. The default value is set at 0.
- `reg_alpha` (alias: alpha): it is the L1 regularization parameter, increasing its value makes the model more conservative. Default is 0.
- `reg_lambda` (alias: lambda): L2 regularization parameter, increasing its value also makes the model conservative. Default is 1.

From: https://towardsdatascience.com/a-guide-to-xgboost-hyperparameters-87980c7f44a9
